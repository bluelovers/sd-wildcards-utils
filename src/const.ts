export const RE_UNSAFE_QUOTE = /['"]/;
export const RE_UNSAFE_QUOTE_DOUBLE = /"/;
export const RE_UNSAFE_VALUE = /^\s*-|[{$~!@}\n|:?#'"%]/;
//export const RE_UNSAFE_PLAIN = /-/;
export const RE_UNSAFE_PLAIN = /-/;
