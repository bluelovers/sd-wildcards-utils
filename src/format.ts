import { IOptionsSharedWildcardsYaml } from './types';

export function stripZeroStr(value: string)
{
	return value
		.replace(/[\x00\u200b]+/g, '')
}

export function trimPrompts(value: string)
{
	return value
		.replace(/\xa0/g, ' ')
		.replace(/^\s+|\s+$/g, '')
		.replace(/^\s+|\s+$/gm, '')
		.replace(/\n\s*\n/g, '\n')
		.replace(/\s+/gm, ' ')
		.replace(/[ ,.]+(?=,|$)/gm, '')
		.replace(/,\s*(?=,|$)/g, '')
		;
}

export function normalizeWildcardsYamlString(value: string)
{
	value = stripZeroStr(value)
		.replace(/\xa0/g, ' ')
		.replace(/[,.]+(?=,)/gm, '')
		.replace(/[ .]+$/gm, '')
		.replace(/\{\s+(\d+(?:\.\d+)?(?:-(?:\d+(?:\.\d+)?)?\$\$|::))/gm, '{$1')
		.replace(/\|\s(\d+(?:\.\d+)?::)/gm, '|$1')
		.replace(/^[ \t]+-[ \t]*$/gm, '')
		.replace(/^([ \t]+-)[ \t][ ,.]+/gm, '$1 ')
		.replace(/^([ \t]+-[^\n]+),+$/gm, '$1')
	;
	return value
}

export function formatPrompts(value: string, opts?: IOptionsSharedWildcardsYaml)
{
	opts ??= {};

	value = stripZeroStr(value);
	value = trimPrompts(value);
	value = normalizeWildcardsYamlString(value);

	if (opts.minifyPrompts)
	{
		value = value
			.replace(/(,)\s+/gm, '$1')
			.replace(/\s+(,)/gm, '$1')
			.replace(/(?<=,\|})\s+/gm, '')
			.replace(/\s+(?=\{(?:\s*\d+(?:\.\d+)?::)?,)/gm, '')
		;
	}

	return value
}

export function stripBlankLines(value: string, appendEOF?: boolean)
{
	value = value
		.replace(/(\r?\n)[\s\r\n\t\xa0]+(\r?\n)/g, '$1$2')
		.replace(/(\r?\n)(?:\r?\n)(?=[\s\t\xa0])/g, '$1')
		.replace(/[ \xa0\t]+$/gm, '')
	;

	if (appendEOF)
	{
		value = value.replace(/\s+$/, '');
		value += '\n\n';
	}

	return value
}
