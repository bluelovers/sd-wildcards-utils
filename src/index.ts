/**
 * Created by user on 2024/5/21.
 */
import { Document, isDocument, ParsedNode, parseDocument, stringify, YAMLMap } from 'yaml';
import {
	defaultOptionsParseDocument,
	defaultOptionsStringify,
	getOptionsFromDocument,
} from './options';
import { _visitNormalizeScalar, visitWildcardsYAML } from './node/node-items';
import { createDefaultVisitWildcardsYAMLOptions, validWildcardsYamlData } from './valid';
import {
	IOptionsParseDocument,
	IOptionsStringify,
	IOptionsVisitor,
	IParseWildcardsYamlInputSource,
	IRecordWildcards,
	IWildcardsYAMLDocument,
	IWildcardsYAMLDocumentParsed,
	IWildcardsYAMLMapRoot,
} from './types';
import { _expandForwardSlashKeys } from './parser';

export * from './util';
export * from './options';
export * from './node/node-items';
export * from './valid';
export * from './prompts/valid-prompts';
export * from './node/node-merge';
export * from './node/node-find';
export * from './prompts/format';
export * from './prompts/prompts';
export * from './check';
export * from './node/node-is';
export type * from './types';

/**
 * Normalizes a YAML document by applying specific rules to its nodes.
 **/
export function normalizeDocument<T extends Document>(doc: T, opts?: IOptionsParseDocument)
{
	let options = getOptionsFromDocument(doc, opts);

	const defaults = createDefaultVisitWildcardsYAMLOptions(options);

	let checkUnsafeQuote = !options.disableUnsafeQuote;

	let visitorOptions: IOptionsVisitor = {
		...defaults,

		Scalar(key, node, parentNodes)
		{
			return _visitNormalizeScalar(key, node, parentNodes, {
				checkUnsafeQuote,
				options,
			})
		},
	};

	visitWildcardsYAML(doc, visitorOptions)
}

/**
 * Converts the given YAML data to a string, applying normalization and formatting.
 *
 * @returns - A string representation of the input YAML data, with normalization and formatting applied.
 *
 * @throws - Throws a `SyntaxError` if the input data is invalid according to the `validWildcardsYamlData` function.
 *
 * @remarks
 * This function takes the input YAML data and applies normalization and formatting using the provided options.
 * If the input data is a `Document` object, it first normalizes the document using the `normalizeDocument` function.
 * Then, it converts the normalized document to a string using the `toString` method with the specified options.
 * If the input data is not a `Document` object, it directly converts the data to a string using the `stringify` function with the specified options.
 *
 * @example
 * ```typescript
 * const yamlData: IRecordWildcards = {
 *   key1: ['value1', 'value2'],
 *   key2: {
 *     subkey1: ['value3', 'value4'],
 *   },
 * };
 *
 * const yamlString = stringifyWildcardsYamlData(yamlData);
 * console.log(yamlString);
 * // Output:
 * // key1:
 * //   - value1
 * //   - value2
 * // key2:
 * //   subkey1:
 * //     - value3
 * //     - value4
 * ```
 */
export function stringifyWildcardsYamlData<T extends IRecordWildcards | IWildcardsYAMLDocument | Document>(data: T | unknown,
	opts?: IOptionsStringify,
)
{
	const isDoc = isDocument(data);

	if (isDoc)
	{
		opts = getOptionsFromDocument(data, opts);
	}

	opts = defaultOptionsStringify(opts);

	if (isDoc)
	{
		normalizeDocument(data, opts);

		return data.toString(opts)
	}

	return stringify(data, opts)
}

/**
 * Parses Stable Diffusion wildcards source to a YAML object.
 *
 * @returns - If `Contents` extends `ParsedNode`, returns a parsed `Document.Parsed` with the specified `Contents` and `Strict`.
 *            Otherwise, returns a parsed `Document` with the specified `Contents` and `Strict`.
 *
 * @throws - Throws a `SyntaxError` if the YAML data is invalid according to the `validWildcardsYamlData` function.
 *
 * @remarks
 * This function parses the given `source` string or Uint8Array to a YAML object.
 * It uses the `parseDocument` function from the `yaml` library with `keepSourceTokens: true` option.
 * Then, it validates the parsed data using the `validWildcardsYamlData` function.
 * Finally, it returns the parsed data.
 */
export function parseWildcardsYaml<Contents extends YAMLMap = IWildcardsYAMLMapRoot, Strict extends boolean = true>(source: IParseWildcardsYamlInputSource,
	opts?: IOptionsParseDocument,
): Contents extends ParsedNode
	? IWildcardsYAMLDocumentParsed<Contents, Strict>
	: IWildcardsYAMLDocument<Contents, Strict>
{
	opts = defaultOptionsParseDocument(opts);

	if (opts.allowEmptyDocument)
	{
		source ??= '';
	}

	let data = parseDocument<Contents, Strict>(source.toString(), opts);

	// expand keys with forward slashes into nested maps
	if (opts.expandForwardSlashKeys)
	{
		_expandForwardSlashKeys(data as Document);
	}

	validWildcardsYamlData(data, opts);

	return data as any
}

export default parseWildcardsYaml

