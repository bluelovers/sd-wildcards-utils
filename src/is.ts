import { isDocument, YAMLMap, Document, isMap } from 'yaml';
import { IWildcardsYAMLDocument, IWildcardsYAMLMapRoot, IWildcardsYAMLPairValue, IWildcardsYAMLScalar } from './types';

export function isWildcardsYAMLDocument<T extends IWildcardsYAMLMapRoot>(node: IWildcardsYAMLDocument<T, true> | Document<T, true>): node is IWildcardsYAMLDocument<T, true>
export function isWildcardsYAMLDocument<T extends IWildcardsYAMLDocument | Document>(doc: any): doc is IWildcardsYAMLDocument
export function isWildcardsYAMLDocument<T extends YAMLMap = IWildcardsYAMLMapRoot>(node: any): node is IWildcardsYAMLDocument<T, true>
export function isWildcardsYAMLDocument(doc: any)
{
	return isDocument(doc)
}

export function isWildcardsYAMLDocumentAndContentsIsMap(doc: any): doc is IWildcardsYAMLDocument
{
	return isDocument(doc) && isMap(doc.contents)
}

export function isWildcardsYAMLMap<K extends IWildcardsYAMLScalar, V extends IWildcardsYAMLPairValue>(doc: IWildcardsYAMLMapRoot<K, V> | YAMLMap.Parsed<K, V> | YAMLMap<K, V>): doc is IWildcardsYAMLMapRoot<K, V>
export function isWildcardsYAMLMap<K extends IWildcardsYAMLScalar = IWildcardsYAMLScalar, V extends IWildcardsYAMLPairValue = IWildcardsYAMLPairValue>(doc: any): doc is IWildcardsYAMLMapRoot<K, V>
export function isWildcardsYAMLMap(doc: any)
{
	return isMap(doc)
}
