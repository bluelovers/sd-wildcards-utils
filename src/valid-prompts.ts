import { ICheckErrorResult } from './types';
import { Extractor, IExtractionError, IExtractionResult, infoNearExtractionError } from '@bluelovers/extract-brackets';
import { _nearString } from './valid';

let _extractor: Extractor;
let _extractor2: Extractor;

export function _handleExtractorError(value: string)
{
	return _handleExtractorErrorCore.bind(null, value)
}

export function _handleExtractorErrorCore(value: string, e: IExtractionError): ICheckErrorResult
{
	if (e)
	{
		let result: IExtractionResult = e.self?.result;

		if (!result)
		{
			return {
				value,
				error: `Invalid Error [UNKNOWN]: ${e}`,
			} satisfies ICheckErrorResult
		}

		let near = infoNearExtractionError(value, e.self)

		return {
			value,
			index: result.index?.[0],
			near,
			error: `Invalid Syntax [BRACKET] ${e.message} near "${near}"`,
		} satisfies ICheckErrorResult
	}
}

export function _checkBracketsCore(value: string, _extractor: Extractor)
{
	return _extractor.extractSync(value, (e) =>
	{
		return _handleExtractorErrorCore(value, e)
	}) as ICheckErrorResult
}

export function _checkBrackets(value: string)
{
	_extractor ??= new Extractor('{', '}');
	return _checkBracketsCore(value, _extractor);
}

export function _checkBrackets2(value: string)
{
	_extractor2 ??= new Extractor('__', '__');
	return _checkBracketsCore(value, _extractor2);
}

export function _checkValue(value: string): ICheckErrorResult
{
	let m = /(?:^|[\s{},])_(?=[^_]|$)|(?<!_)_(?:[\s{},]|$)|\/_+|_+\/(?!\()|\([\w_]+\s*=(?:!|\s*[{}$])/.exec(value)

	if (m)
	{
		let near = _nearString(value, m!.index, m![0]);
		let match = m![0];

		return {
			value,
			match,
			index: m!.index,
			near,
			error: `Invalid Syntax [UNSAFE_SYNTAX] "${match}" in value near "${near}"`,
		}
	}
	else if (/[{}]|__/.test(value))
	{
		return _checkBrackets(value) ?? _checkBrackets2(value)
	}
}
