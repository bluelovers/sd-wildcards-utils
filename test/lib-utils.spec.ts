//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { isWildcardsName, matchDynamicPromptsWildcards, matchDynamicPromptsWildcardsAll } from '../src/util';
import parseWildcardsYaml, { getOptionsFromDocument } from '../src/index';
import { trimPromptsDynamic } from '../src/format';
import { _checkValue } from '../src/valid-prompts';

beforeAll(async () =>
{

});

test(`doc.options`, () => {
	let doc = parseWildcardsYaml(null as any, {
		allowEmptyDocument: true,
	});

	expect(getOptionsFromDocument(doc)).toMatchSnapshot()
});

describe(`matchDynamicPromptsWildcards`, () =>
{

	test.skip(`dummy`, () => {});

	test.each([
		'__season_clothes(season=winter)__',
		'__season_clothes__',
		'__season_clothes(season=__season_clothes__)__',

		'__season_clothes(season=!__season_clothes__)__',

		'__~season_clothes(season=winter)__',

		'__season_clothes(season=__@season_clothes__)__',
		'__season_clothes(season=__~season_clothes__)__',

		'__@season_clothes(season=__season_clothes__)__',
		'__~season_clothes(season=__season_clothes__)__',

		'__season_clothes(season={summer|autumn|winter|spring})__',
		'__season_clothes(season=!{summer|autumn|winter|spring})__',

		'__season_clothes(season={@summer|autumn|winter|spring})__',
		'__season_clothes(season={!summer|autumn|winter|spring})__',

		'__season_clothes(season=!{@summer|autumn|winter|spring})__',
		'__season_clothes(season=!{!summer|autumn|winter|spring})__',

		'__season_clothes(season=)__',

		'__season_clothes(season=)__ ',

		' __season_clothes(season=)__ ',

		' __season_clothes(season= )__ ',

		// ---------------

		'__scope/season_clothes(season=winter)__',
		'__scope/season_clothes__',
		'__scope/season_clothes(season=__scope/season_clothes__)__',

		'__scope/season_clothes(season=!__scope/season_clothes__)__',

		'__~scope/season_clothes(season=winter)__',

		'__scope/season_clothes(season=__@scope/season_clothes__)__',
		'__scope/season_clothes(season=__~scope/season_clothes__)__',

		'__@scope/season_clothes(season=__scope/season_clothes__)__',
		'__~scope/season_clothes(season=__scope/season_clothes__)__',

		'__scope/season_clothes(season={summer|autumn|winter|spring})__',
		'__scope/season_clothes(season=!{summer|autumn|winter|spring})__',

		'__scope/season_clothes(season={@summer|autumn|winter|spring})__',
		'__scope/season_clothes(season={!summer|autumn|winter|spring})__',

		'__scope/season_clothes(season=!{@summer|autumn|winter|spring})__',
		'__scope/season_clothes(season=!{!summer|autumn|winter|spring})__',

		'__scope/season_clothes(season=)__',

		'__scope/season_clothes(season=)__ ',

		' __scope/season_clothes(season=)__ ',

		' __scope/season_clothes(season= )__ ',

	])(`%j`, (input) =>
	{

		let actual = matchDynamicPromptsWildcards(input);

		expect(actual).toMatchSnapshot();

		expect(isWildcardsName(actual.name)).toBeTruthy();

	});

	test.each([
		'__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main__',
	])(`%j`, (input) =>
	{

		let actual = matchDynamicPromptsWildcards(input, {
			unsafe: true,
		});

		expect(actual).toHaveProperty('name');
		expect(actual).toMatchSnapshot();

		actual = matchDynamicPromptsWildcards(input, {
			unsafe: false,
		});

		expect(actual).toBeFalsy();

	});

	test.each([
		'__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main__ __person/regular/haircolor-unconv(k=)__ __lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-*__',
	])(`%j`, (input) =>
	{

		let actual = matchDynamicPromptsWildcardsAll(input, {
			unsafe: true,
		});

		expect(actual).toHaveLength(3);
		expect(actual).toMatchSnapshot();

	});

})

describe(`utils`, () => {

	test.each([
		`\$\{c=!__lazy-wildcards/utils/color-base__\} \n__lazy-wildcards/subject/env-elem/sky_lantern/fn/sky_lantern__`,
		`\$\{v1=!\{\{very |\}small |\}\} \n __lazy-wildcards/subject/costume-ethnicity-breasts/tits-rocket/fn/rocket_tits__`,
		`\$\{person_description = { blond| redhead | brunette}, {green|blue|brown|hazel\} eyes, {tall|average|short}}
A \$\{person_description\} man and a \$\{person_description\} woman`,
	])(`%j`, (input) => {

		let actual = trimPromptsDynamic(input);

		expect(actual).toMatchSnapshot();
	})

});
