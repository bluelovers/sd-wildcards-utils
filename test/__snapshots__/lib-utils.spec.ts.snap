// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`doc.options 1`] = `
{
  "_directives": Directives {
    "atNextDocument": false,
    "docEnd": false,
    "docStart": null,
    "tags": {
      "!!": "tag:yaml.org,2002:",
    },
    "yaml": {
      "explicit": false,
      "version": "1.2",
    },
  },
  "allowEmptyDocument": true,
  "expandForwardSlashKeys": true,
  "intAsBigInt": false,
  "keepSourceTokens": false,
  "logLevel": "warn",
  "prettyErrors": true,
  "strict": true,
  "stringKeys": false,
  "toStringDefaults": {
    "allowMultiRoot": undefined,
    "blockQuote": true,
    "collectionStyle": "block",
    "defaultKeyType": "PLAIN",
    "defaultStringType": "PLAIN",
    "disableUniqueItemValues": undefined,
    "disableUnsafeQuote": undefined,
    "expandForwardSlashKeys": undefined,
    "minifyPrompts": undefined,
    "uniqueKeys": true,
  },
  "uniqueKeys": true,
  "version": "1.2",
}
`;

exports[`matchDynamicPromptsWildcards " __scope/season_clothes(season= )__ " 1`] = `
{
  "isFullMatch": false,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season= )__",
  "variables": "(season= )",
}
`;

exports[`matchDynamicPromptsWildcards " __scope/season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards " __season_clothes(season= )__ " 1`] = `
{
  "isFullMatch": false,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season= )__",
  "variables": "(season= )",
}
`;

exports[`matchDynamicPromptsWildcards " __season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__@scope/season_clothes(season=__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": "@",
  "name": "scope/season_clothes",
  "source": "__@scope/season_clothes(season=__scope/season_clothes__)__",
  "variables": "(season=__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__@season_clothes(season=__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": "@",
  "name": "season_clothes",
  "source": "__@season_clothes(season=__season_clothes__)__",
  "variables": "(season=__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__~scope/season_clothes(season=__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": "~",
  "name": "scope/season_clothes",
  "source": "__~scope/season_clothes(season=__scope/season_clothes__)__",
  "variables": "(season=__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__~scope/season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": "~",
  "name": "scope/season_clothes",
  "source": "__~scope/season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__~season_clothes(season=__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": "~",
  "name": "season_clothes",
  "source": "__~season_clothes(season=__season_clothes__)__",
  "variables": "(season=__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__~season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": "~",
  "name": "season_clothes",
  "source": "__~season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main__ __person/regular/haircolor-unconv(k=)__ __lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-*__" 1`] = `
[
  {
    "isFullMatch": false,
    "isStarWildcards": false,
    "keyword": undefined,
    "name": "lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main",
    "source": "__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main__",
    "variables": undefined,
  },
  {
    "isFullMatch": false,
    "isStarWildcards": false,
    "keyword": undefined,
    "name": "person/regular/haircolor-unconv",
    "source": "__person/regular/haircolor-unconv(k=)__",
    "variables": "(k=)",
  },
  {
    "isFullMatch": false,
    "isStarWildcards": true,
    "keyword": undefined,
    "name": "lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-*",
    "source": "__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-*__",
    "variables": undefined,
  },
]
`;

exports[`matchDynamicPromptsWildcards "__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main",
  "source": "__lazy-wildcards/subject/env-elem-creatures/butterfly/ env-elem-creatures-main__",
  "variables": undefined,
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!__scope/season_clothes__)__",
  "variables": "(season=!__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!{!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!{!summer|autumn|winter|spring})__",
  "variables": "(season=!{!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!{@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!{@summer|autumn|winter|spring})__",
  "variables": "(season=!{@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=!{summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=!{summer|autumn|winter|spring})__",
  "variables": "(season=!{summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=__@scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=__@scope/season_clothes__)__",
  "variables": "(season=__@scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=__~scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=__~scope/season_clothes__)__",
  "variables": "(season=__~scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=__scope/season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=__scope/season_clothes__)__",
  "variables": "(season=__scope/season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season={!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season={!summer|autumn|winter|spring})__",
  "variables": "(season={!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season={@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season={@summer|autumn|winter|spring})__",
  "variables": "(season={@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season={summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season={summer|autumn|winter|spring})__",
  "variables": "(season={summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__scope/season_clothes__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "scope/season_clothes",
  "source": "__scope/season_clothes__",
  "variables": undefined,
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!__season_clothes__)__",
  "variables": "(season=!__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!{!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!{!summer|autumn|winter|spring})__",
  "variables": "(season=!{!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!{@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!{@summer|autumn|winter|spring})__",
  "variables": "(season=!{@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=!{summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=!{summer|autumn|winter|spring})__",
  "variables": "(season=!{summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=)__ " 1`] = `
{
  "isFullMatch": false,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=)__",
  "variables": "(season=)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=__@season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=__@season_clothes__)__",
  "variables": "(season=__@season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=__~season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=__~season_clothes__)__",
  "variables": "(season=__~season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=__season_clothes__)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=__season_clothes__)__",
  "variables": "(season=__season_clothes__)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season={!summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season={!summer|autumn|winter|spring})__",
  "variables": "(season={!summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season={@summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season={@summer|autumn|winter|spring})__",
  "variables": "(season={@summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season={summer|autumn|winter|spring})__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season={summer|autumn|winter|spring})__",
  "variables": "(season={summer|autumn|winter|spring})",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes(season=winter)__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes(season=winter)__",
  "variables": "(season=winter)",
}
`;

exports[`matchDynamicPromptsWildcards "__season_clothes__" 1`] = `
{
  "isFullMatch": true,
  "isStarWildcards": false,
  "keyword": undefined,
  "name": "season_clothes",
  "source": "__season_clothes__",
  "variables": undefined,
}
`;

exports[`utils "\${c=!__lazy-wildcards/utils/color-base__} \\n__lazy-wildcards/subject/env-elem/sky_lantern/fn/sky_lantern__" 1`] = `"\${c=!__lazy-wildcards/utils/color-base__}__lazy-wildcards/subject/env-elem/sky_lantern/fn/sky_lantern__"`;

exports[`utils "\${person_description = { blond| redhead | brunette}, {green|blue|brown|hazel} eyes, {tall|average|short}}\\nA \${person_description} man and a \${person_description} woman" 1`] = `"\${person_description={ blond| redhead | brunette}, {green|blue|brown|hazel} eyes, {tall|average|short}}A \${person_description} man and a \${person_description} woman"`;

exports[`utils "\${v1=!{{very |}small |}} \\n __lazy-wildcards/subject/costume-ethnicity-breasts/tits-rocket/fn/rocket_tits__" 1`] = `"\${v1=!{{very |}small |}}__lazy-wildcards/subject/costume-ethnicity-breasts/tits-rocket/fn/rocket_tits__"`;
